<a href="https://colab.research.google.com/github/doumsdd/goforextract/blob/main/GoForExtract.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Go for Extract
1. Introduction
Dans ce projet, tu vas voir tous les moyens d'extraire des données. Ainsi, tu deviendras le boss de l'étape Extract de l'ETL 👌

2. Le projet
  
  2.1. Extraction d'un fichier csv
Tu l'as déjà fait plusieurs fois mais ça vaut le coup de le refaire car c'est très fréquent de devoir travailler sur des fichiers csv déjà existants.

Le jeu de données qui nous intéresse aujourd'hui met à disposition la base de données brevets de l’INPI. Chaque brevet est géo-localisés à différents niveaux selon l'adresse des déposants et des inventeurs : pays, régions, départements, unités urbaines, communes. Tu peux trouver le fichier csv ici.

Nous te demandons ici d'extraire le fichier csv, de supprimer les quatres dernières colonnes, puis de l'afficher.



In [ ]:
import pandas as pd

# Load the CSV file
df = pd.read_csv("/content/fr-esr-brevets-france-inpi-oeb.csv")

# Display the first few rows of the dataframe
df.head()


Il semble y avoir une erreur lors de la lecture de votre fichier CSV. Le message d'erreur indique qu'il a trouvé une ligne avec un nombre de colonnes différent de ce qui était attendu (29 au lieu de 28), ce qui est généralement le signe d'un problème de formatage dans le fichier.

Nous pouvons essayer de lire le fichier en ignorant les lignes qui posent problème.

 Les données semblent être séparées par des points-virgules (;) plutôt que par des virgules (,). Je vais essayer de lire à nouveau le fichier en utilisant le point-virgule comme séparateur. Ensuite, je supprimerai les quatre dernières colonnes.

In [3]:
import pandas as pd
# Load the CSV file with the correct separator
df = pd.read_csv("/content/fr-esr-brevets-france-inpi-oeb.csv", sep=";", error_bad_lines=False)

# Remove the last 4 columns
df_modified = df.iloc[:, :-4]

# Display the first few rows of the modified dataframe
df_modified.head()


<ipython-input-3-b32cd5eebed8>:3: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  df = pd.read_csv("/content/fr-esr-brevets-france-inpi-oeb.csv", sep=";", error_bad_lines=False)


num_depot origine  annees_publication  annees_delivrance part  \
0   1161944  France                2013             2013.0  Non   
1   0400729  France                2005             2006.0  Non   
2   0512601  France                2007                NaN  Non   
3   1150006  France                2012             2014.0  Non   
4   1101373  France                2012             2014.0  Non   

                        type_d_tablissement  num_siren    id_dep     com_code  \
0                    Grande Entreprise (GE)  632012100     49478        75108   
1                         Personne physique        NaN     12620        35238   
2                         Personne physique        NaN     15837        91477   
3       Personne physique;Personne physique        NaN  489;7548  69034,69034   
4  Entreprise de taille Intermédiaire (ETI)  394342299     37732        94021   

                             com_nom  ...  \
0                          Paris  8e  ...   
1                             Rennes  ...   
2                          Palaiseau  ...   
3  Caluire-et-Cuire,Caluire-et-Cuire  ...   
4                     Chevilly-Larue  ...   

                                      region  code_region_inv  \
0                              Ile-de-France  R11,R11,R11,R11   
1                                   Bretagne              R53   
2                              Ile-de-France              R11   
3  Auvergne-Rhône-Alpes,Auvergne-Rhône-Alpes              R84   
4                              Ile-de-France              R11   

                                          region_inv  \
0  Ile-de-France,Ile-de-France,Ile-de-France,Ile-...   
1                                           Bretagne   
2                                      Ile-de-France   
3                               Auvergne-Rhône-Alpes   
4                                      Ile-de-France   

                               secteur                                 type  \
0                                Privé                      Personne morale   
1                    Personne physique                    Personne physique   
2                    Personne physique                    Personne physique   
3  Personne physique;Personne physique  Personne physique;Personne physique   
4                                Privé                      Personne morale   

                                       etab      domaine_technologique  \
0                    Grande Entreprise (GE)                     Chimie   
1                         Personne physique                     Chimie   
2                         Personne physique  Electronique, électricité   
3       Personne physique;Personne physique                     Autres   
4  Entreprise de taille Intermédiaire (ETI)            Instrumentation   

               sous_domaine_technologique              nom brev_de  
0                   Chimie organique fine          L'oreal     Non  
1                      Chimie alimentaire          delerue     Non  
2  Transmission d’informations numériques         gauthier     Oui  
3                                     BTP  BERNARD;bernard     Non  
4                    Techniques de mesure              Aer     Oui  

[5 rows x 40 columns]

2.2. Extraction d'un fichier json
Direction le Suuuuuud 😎

Sur cette page, tu vas trouver le panorama des festivals en PACA. Télécharge le fichier Json et extrais ses données.
Dis-nous combien il y a de festivals en PACA, quelles sont les 3 villes qui accueillent la majorité des festivals et quel est le mois de l'année le plus chargé en festivals.

In [4]:
# Load the JSON file
df_festivals = pd.read_json("/content/panorama-des-festivals.json")

# Display the first few rows of the dataframe
df_festivals.head()

datasetid                                  recordid  \
0  panorama-des-festivals  f303af5d382cd97c89f233ae9fa84d8f72a7d0c2   
1  panorama-des-festivals  1572da817b780e088c1e81f8da01b5ee16767879   
2  panorama-des-festivals  01111b5088a3adc7e0035d60c75ca4e38ba35cae   
3  panorama-des-festivals  7852d72fc332bed06a81dc800e81c89bcd3d5350   
4  panorama-des-festivals  5d117e289064c0d9b2e661bda9d64a9dfd390ffe   

                                              fields  \
0  {'dept_sk': 13.0, 'coordonnees_insee': [43.550...   
1  {'dept_sk': 13.0, 'coordonnees_insee': [43.299...   
2  {'dept_sk': 13.0, 'coordonnees_insee': [43.299...   
3  {'dept_sk': 83.0, 'coordonnees_insee': [43.136...   
4  {'dept_sk': 13.0, 'coordonnees_insee': [43.293...   

                                            geometry  \
0  {'type': 'Point', 'coordinates': [4.9511813524...   
1  {'type': 'Point', 'coordinates': [5.3822786979...   
2  {'type': 'Point', 'coordinates': [5.3822786979...   
3  {'type': 'Point', 'coordinates': [5.9323963424...   
4  {'type': 'Point', 'coordinates': [5.5633127347...   

                record_timestamp  
0  2019-07-25T13:57:37.191+02:00  
1  2019-07-25T13:57:37.191+02:00  
2  2019-07-25T13:57:37.191+02:00  
3  2019-07-25T13:57:37.191+02:00  
4  2019-07-25T13:57:37.191+02:00

Il semble que le nom de la région n'est pas disponible directement dans les données. Cependant, nous avons le département ('dept_sk') et nous savons que la région PACA comprend les départements suivants :

04 : Alpes-de-Haute-Provence

05 : Hautes-Alpes

06 : Alpes-Maritimes

13 : Bouches-du-Rhône

83 : Var

84 : Vaucluse

Je vais donc utiliser ces informations pour filtrer les festivals qui se déroulent en PACA. Ensuite, j'identifierai les trois villes qui accueillent le plus de festivals, et je déterminerai le mois le plus chargé en festivals.

In [11]:
# Extract information from the 'fields' column and create a new dataframe
df_festivals_details = pd.json_normalize(df_festivals['fields'])

# Define the department codes for PACA region
paca_dept_codes = [4, 5, 6, 13, 83, 84]

# Filter festivals in PACA
df_festivals_paca = df_festivals_details[df_festivals_details['dept_sk'].isin(paca_dept_codes)]


# Display the column names of the dataframe with festival details
df_festivals_details.columns

Index(['dept_sk', 'coordonnees_insee', 'date_de_fin_ancien',
       'mois_indicatif_en_chiffre_y_compris_double_mois', 'nom_departement',
       'departement', 'periodicite', 'mois_habituel_de_debut', 'code_postal',
       'complement_domaine', 'libelle_commune_pour_calcul_cp_insee', 'domaine',
       'date_debut_ancien', 'code_insee', 'commune_principale', 'region',
       'nom_de_la_manifestation', 'ndeg_identification', 'site_web',
       'date_de_creation', 'check_edition', 'ndeg_de_l_edition_2018',
       'ndeg_de_l_edition_2019', 'autres_communes', 'commentaires',
       'soutenu_en_2017_par_le_ministere_de_la_culture',
       'soutenu_en_2017_par_le_centre_national_du_cinema',
       'soutenu_en_2017_par_le_centre_national_du_livre',
       'soutenu_en_2017_par_le_centre_national_des_varietes', 'mois_indicatif',
       'soutenu_en_2018_par_le_ministere_de_la_culture'],
      dtype='object')

Il semble que le nom de la commune est stocké dans la colonne 'libelle_commune_pour_calcul_cp_insee'. Je vais utiliser cette colonne pour identifier les trois villes qui accueillent le plus de festivals. Ensuite, je déterminerai le mois le plus chargé en festivals.

In [14]:
# Count the number of festivals in PACA
num_festivals_paca = df_festivals_paca.shape[0]

# Identify the top 3 cities hosting the most festivals
top_cities = df_festivals_paca['libelle_commune_pour_calcul_cp_insee'].value_counts().nlargest(3)

# Find the busiest month for festivals
busiest_month = df_festivals_paca['mois_indicatif_en_chiffre_y_compris_double_mois'].value_counts().idxmax()

num_festivals_paca, top_cities, busiest_month


(319,
 MARSEILLE 01    56
 CANNES          17
 AVIGNON         16
 Name: libelle_commune_pour_calcul_cp_insee, dtype: int64,
 7.0)

Il y a 319 festivals dans la région PACA (Provence-Alpes-Côte d'Azur).

Les trois villes qui accueillent le plus grand nombre de festivals sont :

Marseille 01 avec 56 festivals
Cannes avec 17 festivals
Avignon avec 16 festivals
Le mois le plus chargé en festivals est le mois de juillet (7ème mois de l'année).

Finies les vacances, retour à la vraie vie 💼💼 : tu enfiles ton costume de banquier pour aller requêter une base de données qui contient tous tes clients.

Télécharge la base de données ici.

Puis tu vas devoir manipuler cette base de données via Python. Cela nécessite l’import de la librairie 'sqlite3'. Replonge dans la ressource et suis les étapes nécessaires pour accéder à la base de données.

Alerte : Les données sont purement fictives, tu pourras même y trouver quelques fantaisies 🙄

Si tu as du temps (tu peux y revenir une fois que tu as testé les 5 manières de collecter de la donnée), voici la liste des requêtes que tu dois réaliser sur cette base de données :

Donner le nom et le prénom de tous les clients.

Il semble qu'il y ait trois tables dans la base de données SQLite : 'client', 'compte' et 'operation'. Puisque nous cherchons les noms et prénoms de tous les clients, nous devons consulter la table 'client'. Je vais maintenant extraire ces informations.

In [16]:
# Import the pandas library
import pandas as pd
import sqlite3
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the names and first names of all clients
client_data = pd.read_sql_query("SELECT nom, prenom FROM client", conn)

# Close the connection
conn.close()

# Display the client data
client_data


nom    prenom
0     Ewing     Grant
1      Buck    Violet
2      Hood  Isabelle
3     Small   Germane
4   Hancock   Vincent
..      ...       ...
95     Koch     Alexa
96   Sparks      Nina
97  Mathews      Elmo
98   Slater   Donovan
99   Sparks     Steel

[100 rows x 2 columns]

Donner le nom et le prénom des clients habitant à Paris.


---



In [17]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the names and first names of clients living in Paris
client_data_paris = pd.read_sql_query("SELECT nom, prenom FROM client WHERE ville = 'Paris'", conn)

# Close the connection
conn.close()

# Display the client data
client_data_paris

nom       prenom
0     Hood     Isabelle
1  Parsons        Leroy
2    Grant     Prescott
3     Hill       Olivia
4      Fry       Sierra
5   Becker  Buckminster

Donner les identifiants des comptes de type Livret A.


In [19]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the first few rows of the 'compte' table
compte_data = pd.read_sql_query("SELECT * FROM compte LIMIT 5", conn)

# Close the connection
conn.close()

# Display the compte data
compte_data


idcompte  idproprietaire                  type
0         1              70  Plan Epargne Actions
1         2              48              Livret A
2         3              87         Assurance Vie
3         4               5        Compte Courant
4         5              63         Assurance Vie

il semble que l'identifiant du compte est stocké dans la colonne 'idcompte'. Je vais utiliser cette colonne pour extraire les identifiants des comptes de type 'Livret A'.

In [20]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the ids of accounts of type 'Livret A'
account_ids = pd.read_sql_query("SELECT idcompte FROM compte WHERE type = 'Livret A'", conn)

# Close the connection
conn.close()

# Display the account ids
account_ids


idcompte
0          2
1         10
2         11
3         13
4         28
..       ...
67       278
68       282
69       295
70       297
71       299

[72 rows x 1 columns]

Il y a donc 72 comptes de type 'Livret A'.

Donner les identifiants des opérations de débit sur le compte d’identifiant égal à 1.


In [21]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the first few rows of the 'operation' table
operation_data = pd.read_sql_query("SELECT * FROM operation LIMIT 5", conn)

# Close the connection
conn.close()

# Display the operation data
operation_data


idop  idcompte     montant informations
0     1         6 -462.192943      Guichet
1     2       269 -288.972643      Guichet
2     3       163 -216.730389     Virement
3     4       236 -298.661908      Guichet
4     5       168  318.978330      Guichet

Il semble que l'identifiant de l'opération est stocké dans la colonne 'idop'. Je vais utiliser cette colonne pour extraire les identifiants des opérations de débit sur le compte avec l'identifiant 1.

In [22]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the ids of debit operations on the account with id 1
operation_ids = pd.read_sql_query("SELECT idop FROM operation WHERE idcompte = 1 AND montant < 0", conn)

# Close the connection
conn.close()

# Display the operation ids
operation_ids


idop
0   836

L'identifiant de l'opération de débit sur le compte avec l'identifiant 1 est 836. Il n'y a qu'une seule opération de débit sur ce compte dans la base de données.

#Donner, sans doublon, les identifiants des propriétaires de livret A, classés par ordre croissant.


In [23]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the ids of owners of 'Livret A' accounts, without duplicates and sorted in ascending order
owner_ids = pd.read_sql_query("SELECT DISTINCT idproprietaire FROM compte WHERE type = 'Livret A' ORDER BY idproprietaire", conn)

# Close the connection
conn.close()

# Display the owner ids
owner_ids


idproprietaire
0                2
1                3
2                5
3                9
4               12
5               13
6               15
7               17
8               22
9               25
10              28
11              29
12              30
13              32
14              33
15              34
16              36
17              38
18              46
19              48
20              49
21              51
22              52
23              54
24              56
25              59
26              60
27              61
28              62
29              63
30              69
31              71
32              72
33              73
34              74
35              78
36              82
37              84
38              87
39              88
40              89
41              91
42              92
43              93
44              95
45              96
46              97
47              98
48             100

#Donner l’identifiant des clients n’ayant pas de livret A.


In [25]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the ids of clients who do not own a 'Livret A' account
client_ids_no_livret_a = pd.read_sql_query("""
SELECT idclient FROM client
WHERE idclient NOT IN (SELECT idproprietaire FROM compte WHERE type = 'Livret A')
""", conn)

# Close the connection
conn.close()

# Display the client ids
client_ids_no_livret_a


idclient
0          1
1          4
2          6
3          7
4          8
5         10
6         11
7         14
8         16
9         18
10        19
11        20
12        21
13        23
14        24
15        26
16        27
17        31
18        35
19        37
20        39
21        40
22        41
23        42
24        43
25        44
26        45
27        47
28        50
29        53
30        55
31        57
32        58
33        64
34        65
35        66
36        67
37        68
38        70
39        75
40        76
41        77
42        79
43        80
44        81
45        83
46        85
47        86
48        90
49        94
50        99

Il y a donc 51 clients qui n'ont pas de compte de type 'Livret A'.

Donner l’identifiant de compte et le type de compte des clients habitant à Paris.


In [26]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the account ids and types for clients living in Paris
account_data_paris = pd.read_sql_query("""
SELECT idcompte, type FROM compte
WHERE idproprietaire IN (SELECT idclient FROM client WHERE ville = 'Paris')
""", conn)

# Close the connection
conn.close()

# Display the account data
account_data_paris


idcompte                  type
0          9         Assurance Vie
1         60        Compte Courant
2         67        Compte Courant
3         77              Livret A
4         82  Plan Epargne Actions
5         91         Assurance Vie
6         98              Livret A
7        113         Assurance Vie
8        114        Compte Courant
9        125              Livret A
10       146         Assurance Vie
11       154         Assurance Vie
12       161         Assurance Vie
13       181              Livret A
14       183        Compte Courant
15       185        Compte Courant
16       200  Plan Epargne Actions
17       204  Plan Epargne Actions
18       219         Assurance Vie
19       233         Assurance Vie
20       234         Assurance Vie
21       242        Compte Courant
22       255        Compte Courant
23       261        Compte Courant
24       263         Assurance Vie
25       264        Compte Courant
26       278              Livret A
27       290         Assurance Vie
28       297              Livret A

Il y a donc 28 comptes appartenant à des clients qui habitent à Paris.

Donner la liste des comptes et les types de compte de Dumbledore.


In [27]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the account ids and types for Dumbledore
account_data_dumbledore = pd.read_sql_query("""
SELECT idcompte, type FROM compte
WHERE idproprietaire IN (SELECT idclient FROM client WHERE nom = 'Dumbledore')
""", conn)

# Close the connection
conn.close()

# Display the account data
account_data_dumbledore


idcompte           type
0        65  Assurance Vie
1        71  Assurance Vie
2       214       Livret A
3       239       Livret A

Il semble que 'Dumbledore' ait 4 comptes : 2 comptes de type 'Assurance Vie' et 2 comptes de type 'Livret A'.

Donner le nombre de clients par ville, classé par ordre alphabétique de villes.


In [28]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the number of clients per city, ordered alphabetically by city
client_counts_per_city = pd.read_sql_query("""
SELECT ville, COUNT(*) as nombre_de_clients FROM client
GROUP BY ville
ORDER BY ville
""", conn)

# Close the connection
conn.close()

# Display the client counts per city
client_counts_per_city


ville  nombre_de_clients
0    Aix-en-Provence                  1
1             Amiens                  2
2             Angers                  2
3           Arcachon                  1
4            Avignon                  1
5            Bayonne                  1
6            Belfort                  2
7           Besançon                  2
8           Blenheim                  1
9            Bologne                  1
10          Bordeaux                  3
11             Brest                  3
12            Calais                  2
13            Cannes                  1
14        Carpentras                  2
15  Clermont-Ferrand                  2
16          Grenoble                  1
17  La Roche-sur-Yon                  1
18              Lens                  1
19             Lille                  4
20          Limanges                  1
21           Limoges                  1
22              Lyon                  5
23             Mande                  1
24         Marseille                  5
25              Metz                  1
26         Montauban                  1
27       Montpellier                  1
28          Mulhouse                  3
29             Nancy                  2
30            Nantes                  3
31              Nice                  4
32            Orange                  1
33           Orléans                  2
34         Palaiseau                  3
35             Paris                  6
36         Perpignan                  1
37          Poudlard                  3
38       Pré-au-Lard                  1
39           Quimper                  1
40            Rangen                  1
41             Reims                  3
42            Rennes                  1
43             Rouen                  2
44         Rue d'Ulm                  1
45     Saint-Etienne                  1
46         Strasboug                  1
47        Strasbourg                  5
48          Toulouse                  2
49       Valencienne                  1
50        Versailles                  1
51                 X                  1

Il semble que Paris soit la ville avec le plus grand nombre de clients (6 clients), suivie de Lyon, Marseille et Strasbourg (5 clients chacune).

Donner la ville ayant le plus de clients.


In [29]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the city with the most clients
city_with_most_clients = pd.read_sql_query("""
SELECT ville, COUNT(*) as nombre_de_clients FROM client
GROUP BY ville
ORDER BY nombre_de_clients DESC
LIMIT 1
""", conn)

# Close the connection
conn.close()

# Display the city with the most clients
city_with_most_clients


ville  nombre_de_clients
0  Paris                  6

La ville qui a le plus de clients est Paris, avec 6 clients.

Trouver le nombre d’opérations effectuées sur chaque compte.

In [30]:
# Reconnect to the SQLite database
conn = sqlite3.connect("/content/py066-ma_banque.db")

# Execute a SQL query to get the number of operations per account
operation_counts_per_account = pd.read_sql_query("""
SELECT idcompte, COUNT(*) as nombre_d_operations FROM operation
GROUP BY idcompte
""", conn)

# Close the connection
conn.close()

# Display the operation counts per account
operation_counts_per_account


idcompte  nombre_d_operations
0           1                    4
1           2                    3
2           3                    4
3           4                    2
4           5                    2
..        ...                  ...
283       296                    4
284       297                    4
285       298                    8
286       299                    3
287       300                    5

[288 rows x 2 columns]

Il y a donc 288 comptes en tout, avec un nombre d'opérations allant de 2 à 8 par compte.